### PR TITLE
packages: add deb support

### PR DESCRIPTION
**Pull request by cmatsuoka**
The Ubuntu deb repository support adds the mechanism to refresh
package lists, fetch and unpack stage packages, and install build
packages.

Packages/repo is planned to be moved into a separate project, where
it will be further refactored. Current code follows the snapcraft
implementation.

Signed-off-by: Claudio Matsuoka <claudio.matsuoka@canonical.com>

- [x] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?

-----

### CODE REVIEW

**Review by cjp256 — APPROVED** [review]

File: craft_parts/packages/base.py [hunk 1]
```diff
@@ -64,7 +64,9 @@ def refresh_build_packages_list(cls) -> None:
 
     @classmethod
     @abc.abstractmethod
-    def install_build_packages(cls, package_names: List[str]) -> List[str]:
+    def install_build_packages(
+        cls, package_names: List[str], list_only: bool = False
```
Comment (cjp256) [inline craft_parts/packages/base.py:68]: The dry-run / list-only functionality should be its own method which install_build_packages() uses.

**Review by cjp256 — APPROVED** [review]

File: craft_parts/packages/base.py [hunk 2]
```diff
@@ -175,7 +177,9 @@ def refresh_build_packages_list(cls) -> None:
         """Refresh the build packages cache."""
 
     @classmethod
-    def install_build_packages(cls, package_names: List[str]) -> List[str]:
+    def install_build_packages(
+        cls, package_names: List[str], list_only: bool = False
+    ) -> List[str]:
         """Install packages on the host system."""
         return []
```
Comment (sergiusens) [inline craft_parts/packages/base.py:184]: this one could also be an abstract method

File: craft_parts/packages/base.py (hunk 1 above)
Comment (cmatsuoka) [reply craft_parts/packages/base.py:68]: I like the idea, I'll add a note for implementation in the big repo refactoring.

File: craft_parts/packages/base.py (hunk 2 above)
Comment (cmatsuoka) [reply craft_parts/packages/base.py:184]: The method in the base class is abstract, this is the concrete implementation in `DummyRepository`.

Comment (sergiusens) [reply craft_parts/packages/base.py:186]: doh, thanks, I missed the collapsed code and thought it was a continuation of what was rendered just above

**Review by sergiusens — APPROVED** [review]

**Pull request merged**